### PR TITLE
ARROW-6541: [Format][C++] Update Columnar.rst for two-part EOS, update C++ implementation

### PR DESCRIPTION
**Pull request by wesm**

### CODE REVIEW

**Review by eerhardt — APPROVED** [review]
Looks good to me.

**Comment by wesm** [comment]
Trying to reproduce the failure that's occurring in Ursabot Python 2.7 build

**Comment by wesm** [comment]
I can repro, not sure why it occurs in Python 2.7 but not 3.6/7

**Comment by wesm** [comment]
@pitrou @emkornfield I found that Python 2.7 wasn't returning aligned memory from `io.BytesIO` and so I am making the metadata read path copy the metadata unconditionally if it is unaligned (even if the IPC format is implemented properly)

**Review by BryanCutler — APPROVED** [review]
LGTM

File: cpp/src/arrow/ipc/metadata_internal.h [hunk 1]
```diff
@@ -58,6 +58,9 @@ class DictionaryMemo;
 
 namespace internal {
 
+// This 0xFFFFFFFF value is the first 4 bytes of a valid IPC message
+constexpr int32_t kIpcContinuationToken = -1;
```
Comment (emkornfield) [inline cpp/src/arrow/ipc/metadata_internal.h]: for some reason i thought there were linkage issues in C++11 using constexpr in a header.

**Review by emkornfield — APPROVED** [review]
Looks OK to me, feel free to merge when build passes.

**Pull request closed without merging**